Report item routing in response payload - feedback needed

tlosbo opened the pull request:
This PR is an attempt at updating the /api/reports payload response for this check list item in issue #739 
> Should warn if data is going nowhere. eg, a warning that says "14/35 items will not go to any destination". Should list the tracking_ids of items that went nowhere - maybe limit size of this list to first 25??. ALTERNATIVE, OR IN ADDITION: Maybe we should have a verbose response mode, where we give the status of every single row - where did it go or not go?

## Changes
- added two new elements to the response: orphanItemCount and OrphanItemIds similar to errors and warnings (open to suggestions on the name, was going with the SQL orphaned records concept)
```
  "orphanItemCount" : 0,
  "orphanItemsIds" : [ ],
```
- added a conditional option `ShowReportRouting` that lists each tracking id and the code of the receiver where the item was sent
```
  "reportRouting" : [ {
    "trackingId" : "651477",
    "destinations" : [ "al-phd" ]
  }, {
    "trackingId" : "775659",
    "destinations" : [ "ak-phd" ]
  }, {
    "trackingId" : "382382",
    "destinations" : [ "az-phd", "pima-az-phd" ]
  } ],
```

## Questions / thoughts / suggestions for these changes
- Separate count/array elements vs having an entry in the warnings array?
- I added the options element for the report routing because of the verboseness of the response. It also contains a lot of duplicate data that is in the destinations element, these could be combined. For large reports, we can end up with very large payload responses. Should we use different query string parameters to control the response payload? i.e. `showWarnings=true/false`, `showDestinations=true/false` etc.?
- What is the best way to write a unit test for this with most of the functions being private? expose a method that calls the private functions needed for explicit unit testing?

## Checklist

### Testing
- [x] Tested locally?
- [x] Ran `quickTest all`?
- [ ] Ran `./prime test` against local Docker ReportStream container?
- [ ] Downloaded a file from `http://localhost:7071/api/download`?
- [ ] Added tests?

### Security
- [x] Did you check for sensitive data, and remove any? 
- [x] Does logging contain sensitive data?  
- [ ] Are additional approvals needed for this change?
- [ ] Are there potential vulnerabilities or licensing issues with any new dependencies introduced?

### Process
- [ ] Includes a summary of what a code reviewer should verify?
- [x] Updated the release notes?
Will need to add release notes after feedback
- [ ] Database changes are submitted as a separate PR?
- [ ] DevOps team has been notified if PR requires ops support?

## Security
*List potential security threats and mitigations in the PR*

## Fixes
*List GitHub issues this PR fixes*
- #739

## To Be Done
*Create GitHub issues to track the work remaining, if any*
- #issue

jimduff-usds reviewed (commented):
Thanks @tlosbo !  

I have not looked at the code yet, but I do have a couple questions and comments first - 

1.  I wonder if `orphaned` will be confusing, or evocative.   I had also used the word withered in other places, which is probably worse.  Stunted?  Stuck?  Undigestable?   Data Impaction?  @drew-usds - Help!!  in our json response back to senders, what's a brief way we can describe data that didn't go to any states - data that came in but never went out?  (Our current wording is `orphanedItems`)

One solution here would be to have an extra destination called "nowhere", following the same format as our other destinations.  Here, for example, 3 items went to AZ, and 2 went nowhere:

  "destinations" : [ {
    "organization" : "Arizona PHD",
    "organization_id" : "az-phd",
    "service" : "elr-prod",
    "sending_at" : "immediately",
    "itemCount" : 3
  }, {
    "organization" : "Data that did not go to any state/local jurisdiction",
    "organization_id" : "ningun-lado,
    "service" : "no-where",
    "sending_at" : "never",
    "itemCount" : 2
  }, {

How much of a pain would it be to implement that way?  @drew-usds , any naming thoughts here, with this idea?

2.  @tlosbo , it would be super helpful if can you give an example of a _whole_ json response, either here or in the ticket. Then we can discuss.

3.   Uh, unfortunately I didn't think of this sooner, but:    Soon much our data will be going to HHSProtect, which will render this feature kinda useless.   So I propose that data that only goes to receivers whose organization is `jurisdiction: FEDERAL` are considered orphaned as well.   Right now we only have one receiver in that category, HHSProtect, but we might have other Federal receivers in the future.    _So, I propose that if data ONLY goes to FEDERAL jurisdictions, that's the same as having gone no-where._  You'll see that in the way I labelled things, above.  (I suppose we could even break this by Federal, State, then Local, but that seems like information overload for the Sender)

@tlosbo , how much of a pain-in-the-neck would it be to implement it that way?

jimduff-usds reviewed (commented):
One more thought - maybe rather than ShowReportRouting, just have a more generic

Verbose

I'd like a way to cut off the number of warnings, and that might fall under Verbose as well.   Kills two logorrhea's with one stone.

tlosbo commented:
```
% curl -X POST -H 'client: simple_report' \
  -H 'Content-Type: text/csv' \
  --data-binary '@./examples/upload-fake-data/Test.csv' \
  'http://localhost:7071/api/reports?option=ShowReportRouting'
```
is producing this JSON response locally
```
{
  "id": "cd53aeae-36df-4510-ba59-99a6019e4796",
  "timestamp": "2021-06-03T21:14:23.895978Z",
  "topic": "covid-19",
  "reportItemCount": 23,
  "orphanItemCount": 1,
  "orphanItems": [
    "037746"
  ],
  "destinations": [
    {
      "organization": "Arizona PHD",
      "organization_id": "az-phd",
      "service": "elr-prod",
      "sending_at": "immediately",
      "itemCount": 1
    },
    {
      "organization": "Pima County, Arizona PHD",
      "organization_id": "pima-az-phd",
      "service": "elr",
      "sending_at": "immediately",
      "itemCount": 1
    },
    {
      "organization": "Colorado Department of Public Health and Environment",
      "organization_id": "co-phd",
      "service": "elr",
      "sending_at": "immediately",
      "itemCount": 1
    },
    {
      "organization": "Florida Department of Health",
      "organization_id": "fl-phd",
      "service": "elr",
      "sending_at": "immediately",
      "itemCount": 1
    },
    {
      "organization": "Louisiana Department of Health",
      "organization_id": "la-doh",
      "service": "elr",
      "sending_at": "immediately",
      "itemCount": 1
    },
    {
      "organization": "Delaware Division of Public Health",
      "organization_id": "de-dph",
      "service": "elr",
      "sending_at": "immediately",
      "itemCount": 1
    },
    {
      "organization": "California Department of Public Health",
      "organization_id": "ca-dph",
      "service": "elr",
      "sending_at": "immediately",
      "itemCount": 1
    },
    {
      "organization": "Massachusetts Public Health Department",
      "organization_id": "ma-phd",
      "service": "elr",
      "sending_at": "immediately",
      "itemCount": 1
    },
    {
      "organization": "Alabama Public Health Department",
      "organization_id": "al-phd",
      "service": "elr",
      "sending_at": "immediately",
      "itemCount": 1
    },
    {
      "organization": "Michigan Public Health Department",
      "organization_id": "mi-phd",
      "service": "elr",
      "sending_at": "immediately",
      "itemCount": 1
    },
    {
      "organization": "Maryland",
      "organization_id": "md-doh",
      "service": "elr",
      "sending_at": "immediately",
      "itemCount": 1
    },
    {
      "organization": "Maryland Public Health Department",
      "organization_id": "md-phd",
      "service": "elr",
      "sending_at": "immediately",
      "itemCount": 1
    },
    {
      "organization": "Alaska Public Health Department",
      "organization_id": "ak-phd",
      "service": "elr",
      "sending_at": "immediately",
      "itemCount": 1
    },
    {
      "organization": "Arkansas Public Health Department",
      "organization_id": "ar-phd",
      "service": "elr",
      "sending_at": "immediately",
      "itemCount": 1
    },
    {
      "organization": "Connecticut Public Health Department",
      "organization_id": "ct-phd",
      "service": "elr",
      "sending_at": "immediately",
      "itemCount": 1
    },
    {
      "organization": "District of Columbia Public Health Department",
      "organization_id": "dc-phd",
      "service": "elr",
      "sending_at": "immediately",
      "itemCount": 1
    },
    {
      "organization": "Georgia Public Health Department",
      "organization_id": "ga-phd",
      "service": "elr",
      "sending_at": "immediately",
      "itemCount": 1
    },
    {
      "organization": "Hawaii Public Health Department",
      "organization_id": "hi-phd",
      "service": "elr",
      "sending_at": "immediately",
      "itemCount": 1
    },
    {
      "organization": "Idaho Public Health Department",
      "organization_id": "id-phd",
      "service": "elr",
      "sending_at": "immediately",
      "itemCount": 1
    },
    {
      "organization": "Illinois Public Health Department",
      "organization_id": "il-phd",
      "service": "elr",
      "sending_at": "immediately",
      "itemCount": 1
    },
    {
      "organization": "Indiana Public Health Department",
      "organization_id": "in-phd",
      "service": "elr",
      "sending_at": "immediately",
      "itemCount": 1
    },
    {
      "organization": "Iowa Public Health Department",
      "organization_id": "ia-phd",
      "service": "elr",
      "sending_at": "immediately",
      "itemCount": 1
    },
    {
      "organization": "Kansas Public Health Department",
      "organization_id": "ks-phd",
      "service": "elr",
      "sending_at": "immediately",
      "itemCount": 1
    },
    {
      "organization": "Maine Public Health Department",
      "organization_id": "me-phd",
      "service": "elr",
      "sending_at": "immediately",
      "itemCount": 1
    }
  ],
  "destinationCount": 24,
  "reportRouting": [
    {
      "651477": [
        "al-phd"
      ]
    },
    {
      "775659": [
        "ak-phd"
      ]
    },
    {
      "382382": [
        "az-phd",
        "pima-az-phd"
      ]
    },
    {
      "150986": [
        "ar-phd"
      ]
    },
    {
      "897255": [
        "ca-dph"
      ]
    },
    {
      "587326": [
        "co-phd"
      ]
    },
    {
      "361557": [
        "ct-phd"
      ]
    },
    {
      "027422": [
        "de-dph"
      ]
    },
    {
      "461715": [
        "dc-phd"
      ]
    },
    {
      "438910": [
        "fl-phd"
      ]
    },
    {
      "530084": [
        "ga-phd"
      ]
    },
    {
      "073713": [
        "hi-phd"
      ]
    },
    {
      "444035": [
        "id-phd"
      ]
    },
    {
      "846060": [
        "il-phd"
      ]
    },
    {
      "588506": [
        "in-phd"
      ]
    },
    {
      "761325": [
        "ia-phd"
      ]
    },
    {
      "453601": [
        "ks-phd"
      ]
    },
    {
      "037746": []
    },
    {
      "079098": [
        "la-doh"
      ]
    },
    {
      "162390": [
        "me-phd"
      ]
    },
    {
      "250196": [
        "md-doh",
        "md-phd"
      ]
    },
    {
      "028707": [
        "ma-phd"
      ]
    },
    {
      "816784": [
        "mi-phd"
      ]
    }
  ],
  "warningCount": 0,
  "errorCount": 0,
  "errors": [],
  "warnings": []
}
```
The reportRouting is using the tracking element identifier to show which receiver orgs the the item routed too. I agree this can be combined in the destinations.

I like the idea of ```Verbose``` mode where the warnings can be truncated or capped and something around the routing or destinations.

tlosbo commented:
Regarding implementing your comment around items only going to jurisdiction federal, I think I can implement that without too much effort. I'll create a map of the receiver orgs and their jurisdiction as part of the function so it can be used as a point of reference when building the JSON response.

MauriceReeves-usds commented:
@tlosbo @jimduff-usds I like the idea of a verbosity flag for the query string as well, with some kind of default to less verbose. I was also thinking that for things like the orphan count and IDs, etc we should just add some REST apis for that kind of stuff:

`/report/cd53aeae-36df-4510-ba59-99a6019e4796/orphans`
and
`/report/cd53aeae-36df-4510-ba59-99a6019e4796/routing`

Which would allow us to get more detail about a report and what happened to it.

I also think it would be good to add something for individual items which would show us where an item ended up:
`/report/cd53aeae-36df-4510-ba59-99a6019e4796/item/<TRACKING ID>/`

It could look something like
```json
{
    itemTrackingId: "TRACKING ID",
    isOrphaned: false,
    fromReportId: "cd53aeae-36df-4510-ba59-99a6019e4796",
    sender: "default",
    destinations: [
        { reportId: "<DESTINATION REPORT ID>", receiver: "HHS Protect", sentOn: <DATE SENT> },
        ....
    ]
}
```

This could be really useful for our purposes, especially during customer support when we want to know where the data came from and where it went and when.

Finally @tlosbo as for testing all of this, @carlosfelix2 and I have been discussing adding a new layer of testing locally that would involve spinning up Docker and submitting requests and consuming the responses. We can automate this and build it into gradle so we have some more advanced integration testing. It's an option. The biggest thing blocking it is #1012, which would install the docker compose plugin into gradle.

tlosbo commented on prime-router/docs/openapi.yml:
```diff
@@ -471,6 +471,9 @@ components:
             type: string
+        destinationCount:
+          description: total destinations recieving the report(s)
+          type: integer
```
Noticed this was missing from the yml.

tlosbo commented:
@jimduff-usds @MauriceReeves-usds:
I removed the orphan count/items from the payload and modified the report routing mapping to allow for jurisdiction checks based on the `FEDERAL`, `STATE` and `COUNTY` level. I m building a list of the submitted items with the tracking id and a map keyed by jurisdiction with a list of the receiver organization names:
```
[
  {
    "trackingId": 960022,
    "destinations": {
      "STATE": ["al-phd"],
      "FEDERAL": ["hhsprotect"]
    }
  },
  {
    "trackingId": 882911,
    "destinations": {
      "FEDERAL": ["hhsprotect"]
    }
  },
  {
    "trackingId": 314448,
    "destinations": {
      "STATE": ["az-phd"],
      "FEDERAL": ["hhsprotect"]
    }
  },
  {
    "trackingId": 477507,
    "destinations": {
      "FEDERAL": ["hhsprotect"],
      "STATE:": ["ar-phd"]
    }
  },
  {
    "trackingId": 143171,
    "destinations": {
      "FEDERAL": ["hhsprotect"]
    }
  }
]
```
When building the API response, any trackingIds with no destinations or only federal destinations are filtered out and used to create the "nowhere" destination with an output similar to what @jimduff-usds suggested. Here's an example of the JSON response with 5 entries and 2 matching `FEDERAL` only:
```
{
  "id" : "866e0700-bf99-4704-a0d3-9c59523080a7",
  "timestamp" : "2021-06-10T13:40:39.399943Z",
  "topic" : "covid-19",
  "reportItemCount" : 5,
  "destinations" : [ {
    "organization" : "HHSProtect",
    "organization_id" : "hhsprotect",
    "service" : "elr",
    "sending_at" : "2021-06-10T09:41-04:00",
    "itemCount" : 5
  }, {
    "organization" : "Arizona PHD",
    "organization_id" : "az-phd",
    "service" : "elr-prod",
    "sending_at" : "2021-06-10T06:41-07:00",
    "itemCount" : 1
  }, {
    "organization" : "Alabama Public Health Department",
    "organization_id" : "al-phd",
    "service" : "elr",
    "sending_at" : "2021-06-10T09:41-04:00",
    "itemCount" : 1
  }, {
    "organization" : "Arkansas Public Health Department",
    "organization_id" : "ar-phd",
    "service" : "elr",
    "sending_at" : "2021-06-10T09:41-04:00",
    "itemCount" : 1
  }, {
    "organization" : "Data not routing to any state/local jurisdiction",
    "organization_id" : "nowhere-org",
    "service" : "nowhere-receiver",
    "sending_at" : "immediately",
    "itemCount" : 2
  } ],
  "destinationCount" : 5,
  "warningCount" : 0,
  "errorCount" : 0,
  "errors" : [ ],
  "warnings" : [ ]
}
```
Currently the sending_at is "immediately" but Pablo has some changes to be merged in that will change that to "never".

jimduff-usds reviewed (commented):
Nice.

I'm still back and forth about whether to put the detailed tracking ids in the destinations section, or keep it separate.   I guess since its separate, leave it there for now.

I agree with @mauricereeves  that an api that gives back all kind of detailed descendant info is a great idea.  This is the same as the Audit Trail API we've been talking about for months.  

@tlosbo, I think this is ready, but I'm going to hold off on merging until the verbose mode is ready -- this stuff is wonderful but very verbose!

It would be cool if the brief summary of what went no-where was in the regular mode:
That is, this:
```
{
    "organization" : "Data not routing to any state/local jurisdiction",
    "organization_id" : "nowhere-org",
    "service" : "nowhere-receiver",
    "sending_at" : "immediately",
    "itemCount" : 2
  } 
```
always shows up in the response, even if not in verbose mode.

jimduff-usds reviewed (commented):
I believe @tlosbo plans to remove this PR and include these changes in another PR.

tlosbo commented:
I think I will try to rebase first before rolling these into another PR. I thought it was tightly coupled with the error/warning messaging, but after looking at the changes it may be easier to get this out quicker as it's own.

tlosbo commented:
@jimduff-usds I have resolved the merge conflicts and would like to have this reviewed again. Adding it to the other PR didn't make as much sense after looking at it. Couple of outstanding questions to resolve:

1. do we want to include the item ids/rows that did not get sent?
2. The very verbose report routing in comment https://github.com/CDCgov/prime-reportstream/pull/1231#issuecomment-856205758 : given that we are trying to condense warnings, this is probably too much. Do we want an option to show it or persist to the DB at a later point.

tlosbo commented on prime-router/src/main/kotlin/Schema.kt:
```diff
@@ -79,6 +79,16 @@ data class Schema(
+    fun getTrackingId(row: Int, trackingId: String? = null): String {
+        return trackingId ?: "row$row"
+    }
+
```
Had to implement a default trackingId. Without this, any schema with no tracking element returns null and I have no way to link items to outgoing reports.

tlosbo commented on prime-router/src/main/kotlin/azure/ReportFunction.kt:
```diff
@@ -30,7 +31,8 @@ private const val DEFAULT_PARAMETER = "default"
 private const val ROUTE_TO_SEPARATOR = ","
-
+private const val VERBOSE_PARAMETER = "verbose"
+private const val VERBOSE_ALL = "all"
```
This will also be used with the condensed error/warning messaging.

tlosbo commented on prime-router/src/main/kotlin/azure/ReportFunction.kt:
```diff
@@ -395,8 +409,26 @@ class ReportFunction {
+                    it.writeEndObject()
+                }
+            }
+            it.writeEndArray()
```
This produces this Json output:
```json
"routing" : [ {
    "trackingId" : "651477",
    "destinations" : [ ]
  }, {
    "trackingId" : "775659",
    "destinations" : [ ]
  }, {
    "trackingId" : "382382",
    "destinations" : [ ]
}]
```

tlosbo commented:
@jimduff-usds This is ready to review now. I need to update the openapi.yml file for the new routing element in the response and the destinations needs to me updated as it is not as verbose.

Without the `verbose` param, the routing will be empty. if you pass a query string verbose=all it will show the routing like this:
```json
"routing" : [ {
    "trackingId" : "651477",
    "destinations" : [ ]
  }, {
    "trackingId" : "444035",
    "destinations" : [ "id-phd.elr" ]
  }, {
    "trackingId" : "846060",
    "destinations" : [ ]
  }, {
    "trackingId" : "162390",
    "destinations" : [ "me-phd.elr" ]
  }, {
    "trackingId" : "250196",
    "destinations" : [ "md-doh.elr", "md-phd.elr" ]
  }, {
    "trackingId" : "816784",
    "destinations" : [ ]
  } ]
```

tlosbo commented:
@jimduff-usds per our slack conversation, this is now ready to review. The openapi.yml was updated and the default trackingId has been modified from row to reportIndex.

jimduff-usds commented on prime-router/src/main/kotlin/azure/ActionHistory.kt:
```diff
@@ -28,6 +28,8 @@ import org.jooq.impl.DSL
 import java.io.ByteArrayOutputStream
 import java.time.OffsetDateTime
 
+private const val NOWHERE_ORG = "nowhere-org"
```
```suggestion
private const val NOWHERE_ORG = "n/a"
```

jimduff-usds commented on prime-router/src/main/kotlin/azure/ActionHistory.kt:
```diff
@@ -514,10 +538,24 @@ class ActionHistory {
+                jsonGen,
+                "Data not routing to any state/local jurisdiction",
+                NOWHERE_ORG,
+                "nowhere-receiver",
```
```suggestion
                "n/a",
```

jimduff-usds commented on prime-router/src/main/kotlin/azure/ActionHistory.kt:
```diff
@@ -514,10 +538,24 @@ class ActionHistory {
+        if (noWhereItems.isNotEmpty()) {
+            prettyPrintDestinationJson(
+                jsonGen,
+                "Data not routing to any state/local jurisdiction",
```
```suggestion
                "Note: some data is not routing to any state/local jurisdiction",
```

jimduff-usds reviewed (approved):
wording changes proposed.

jimduff-usds commented on prime-router/src/main/kotlin/azure/ActionHistory.kt:
```diff
@@ -514,10 +538,24 @@ class ActionHistory {
+                "Note: some data is not routing to any state/local jurisdiction",
+                NOWHERE_ORG,
+                "n/a",
+                "never - no matching receivers ",
```
```suggestion
                "n/a",
```

jimduff-usds reviewed (changes requested):
Sigh.  See #1474

tlosbo commented:
Closing this and resubmitting with changes based on comments here and in #1474
